internal/storage/serde.go

```diff
@@ -417,3 +664,362 @@ func NewBinlogDeserializeReader(blobs []*Blob, PKfieldID UniqueID) (*Deserialize
+	}
+	return &singleFieldRecordWriter{
+		fw:      fw,
+		fieldId: fieldId,
+	}, nil
+}
+
+type SerializeWriter[T any] struct {
+	rw         RecordWriter
+	serializer Serializer[T]
+	batchSize  int
+
+	buffer            []T
+	pos               int
+	writtenMemorySize uint64
+}
+
+func (sw *SerializeWriter[T]) Flush() error {
+	buf := sw.buffer[:sw.pos]
+	r, size, err := sw.serializer(buf)
+	if err != nil {
+		return err
+	}
+	if err := sw.rw.Write(r); err != nil {
+		return err
+	}
+	r.Release()
+	sw.pos = 0
+	sw.writtenMemorySize += size
+	return nil
+}
+
+func (sw *SerializeWriter[T]) Write(value T) error {
+	if sw.buffer == nil {
+		sw.buffer = make([]T, sw.batchSize)
+	}
+	sw.buffer[sw.pos] = value
+	sw.pos++
+	if sw.pos == sw.batchSize {
+		if err := sw.Flush(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (sw *SerializeWriter[T]) WrittenMemorySize() uint64 {
+	return sw.writtenMemorySize
+}
+
+func (sw *SerializeWriter[T]) Close() error {
+	sw.rw.Close()
+	return sw.Flush()
+}
+
+func NewSerializeRecordWriter[T any](rw RecordWriter, serializer Serializer[T], batchSize int) *SerializeWriter[T] {
+	return &SerializeWriter[T]{
+		rw:         rw,
+		serializer: serializer,
+		batchSize:  batchSize,
+	}
+}
+
+type simpleArrowRecord struct {
+	Record
+
+	r      arrow.Record
+	schema map[FieldID]schemapb.DataType
+
+	field2Col map[FieldID]int
+}
+
+func (sr *simpleArrowRecord) Schema() map[FieldID]schemapb.DataType {
+	return sr.schema
+}
+
+func (sr *simpleArrowRecord) Column(i FieldID) arrow.Array {
+	colIdx, ok := sr.field2Col[i]
+	if !ok {
+		panic("no such field")
+	}
+	return sr.r.Column(colIdx)
+}
+
+func (sr *simpleArrowRecord) Len() int {
+	return int(sr.r.NumRows())
+}
+
+func (sr *simpleArrowRecord) Release() {
+	sr.r.Release()
+}
+
+func newSimpleArrowRecord(r arrow.Record, schema map[FieldID]schemapb.DataType, field2Col map[FieldID]int) *simpleArrowRecord {
+	return &simpleArrowRecord{
+		r:         r,
+		schema:    schema,
+		field2Col: field2Col,
+	}
+}
+
+type BinlogStreamWriter struct {
+	collectionID UniqueID
+	partitionID  UniqueID
+	segmentID    UniqueID
+	fieldSchema  *schemapb.FieldSchema
+
+	memorySize int // To be updated on the fly
+
+	buf bytes.Buffer
+	rw  RecordWriter
+}
+
+func (bsw *BinlogStreamWriter) GetRecordWriter() (RecordWriter, error) {
+	if bsw.rw != nil {
+		return bsw.rw, nil
+	}
+
+	fid := bsw.fieldSchema.FieldID
+	dim, _ := typeutil.GetDim(bsw.fieldSchema)
+	rw, err := newSingleFieldRecordWriter(fid, arrow.Field{
+		Name: strconv.Itoa(int(fid)),
+		Type: serdeMap[bsw.fieldSchema.DataType].arrowType(int(dim)),
+	}, &bsw.buf)
+	if err != nil {
+		return nil, err
+	}
+	bsw.rw = rw
+	return rw, nil
+}
+
+func (bsw *BinlogStreamWriter) Finalize() (*Blob, error) {
+	if bsw.rw == nil {
+		return nil, io.ErrUnexpectedEOF
+	}
+	bsw.rw.Close()
+
+	var b bytes.Buffer
+	if err := bsw.writeBinlogHeaders(&b); err != nil {
+		return nil, err
+	}
+	if _, err := b.Write(bsw.buf.Bytes()); err != nil {
```

Comment:
According to the doc of `buffer.Writer`, err will always be nil and might be panic with ErrTooLarge. Please remove the err check here,  might need to capture panic here.
![image](https://github.com/milvus-io/milvus/assets/51370125/b67a7055-2efb-4033-8061-ac29f05daea1)

Reply:
Thanks for the comment!

Well, the redundant error check may not be an issue. As for the `ErrTooLarge` panic, I guess we could ignore it as well, since it is thrown only if the writing length reaches max int.